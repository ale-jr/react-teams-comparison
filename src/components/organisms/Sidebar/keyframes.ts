import { keyframes } from 'styled-components'

export const slideIn = keyframes`
  0% {
    width: 0%;
    overflow: hidden;
  }
  100% {
    width: 224px;
  }
`
export const slideOut = keyframes`
  0% {
    width: 224px;
  }
 100% {
    width: 0;
  }
`

export const ulSlideIn = keyframes`
    0%{
        transform: translateX(-100%);
    }

    100%{
        transform: translateX(0%);
    }
`

export const ulSlideOut = keyframes`
    0%{
        transform: translateX(0%);
    }

    100%{
        transform: translateX(-100%);
    }
`

export const backdropFadeIn = keyframes`
    0%{
        opacity: 0;
    }

    99%{
        opacity: 1;
    }
`

export const backdropFadeOut = keyframes`
    0%{
        opacity: 1;
    }

    99.99%{
        width: 100vw;
        opacity: 0;
    }

    100%{
        width:0;
    }
`

