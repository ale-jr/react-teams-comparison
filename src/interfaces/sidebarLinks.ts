import { IconDefinition } from "@fortawesome/fontawesome-svg-core"
export interface SidebarLink {
    icon?: IconDefinition
    title: string
    to: string
    links?: SidebarLink[]
}