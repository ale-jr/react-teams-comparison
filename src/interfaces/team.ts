export interface TeamValues {
  participation: number;
  eNPS: number;
  overallEngagement: number;
  recognition: number;
  feedback: number;
  relationshipWithPeers: number;
  relationshipWithManager: number;
  satisfaction: number;
  alignment: number;
  happiness: number;
  wellness: number;
  personalGrowth: number;
  ambassadorship: number;
}

export interface Team extends TeamValues {
  name: string;
}
