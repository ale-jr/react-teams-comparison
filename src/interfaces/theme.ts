export interface ColorVariants {
  darkest?: string;
  dark?: string;
  default: string;
  light: string;
  lightest?: string;
  contrast: string;
  "0"?: string;
  "50"?: string;
  "100"?: string;
  "200"?: string;
  "300"?: string;
  "400"?: string;
  "500"?: string;
  "600"?: string;
}

export type ColorVariantKeys = keyof ColorVariants;

export interface Sizing {
  xs: string;
  sm: string;
  md: string;
  lg: string;
  xl: string;
}

export type SizingKeys = keyof Sizing;

export interface Typography {
  size: Sizing;
  weight: {
    regular: string;
    semibold: string;
    bold: string;
  };
  font: {
    heading: string;
    body: string;
  };
}

export interface Colors {
  neutral: ColorVariants;
  primary: ColorVariants;
  success: ColorVariants;
  danger: ColorVariants;
}
export type ColorKeys = keyof Colors;

export interface Animation {
  duration: string;
  timingFunction: string;
  shorthand: string;
}

export interface Theme {
  colors: Colors;
  spacing: Sizing;
  typography: Typography;
  animation: Animation;
}
