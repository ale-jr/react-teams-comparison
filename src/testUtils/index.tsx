import { render, RenderOptions, RenderResult } from '@testing-library/react'
import ThemeProvider from 'providers/ThemeProvider'
import { ReactElement } from 'react'


const Providers: React.FC<{
    children: JSX.Element | JSX.Element[]
}> = ({ children }) => <ThemeProvider>{children}</ThemeProvider>


const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "queries">): RenderResult => render(ui, {
    wrapper: Providers,
    ...options
})

export * from '@testing-library/react'
export {customRender as render}