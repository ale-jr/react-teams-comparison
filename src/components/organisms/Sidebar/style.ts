import styled from 'styled-components'
import { backdropFadeIn, backdropFadeOut, slideIn, slideOut, ulSlideIn, ulSlideOut, } from './keyframes'

interface SidebarContainerProps {
    $collapsed: boolean
    $preventAnimation: boolean
}

export const SidebarContainer = styled.nav<SidebarContainerProps>`
    //Position and Layout


    //Display and visibility


    //Clipping


    //Animation
    animation: ${({ $collapsed }) => $collapsed ? slideOut : slideIn} ${({ $preventAnimation }) => $preventAnimation ? '0ms' : '200ms'} linear both;
    
    //Box model (From outside in)
    height: 100%;
    border-right: 1px solid ${({ theme }) => theme.colors.neutral["200"]};
    padding-top: ${({ theme }) => theme.spacing.sm};

    //Background
    background-color: ${({ theme }) => theme.colors.neutral['0']};

    //Typography


    //Pseudo-classes and pseudo-elements

    ul{
        padding: 0;
    }

    ul:first-child{
        animation: ${({ $collapsed }) => $collapsed ? ulSlideOut : ulSlideIn} ${({ $preventAnimation }) => $preventAnimation ? '0ms' : '200ms'} linear both;

        position: absolute;
    }

    @media only screen and (max-width: 425px) {
        position: absolute;
        z-index: 98;

        height: calc(100% - 4.875rem);

    }
`

interface BackdropProps {
    $collapsed: boolean
}

export const Backdrop = styled.div<BackdropProps>`
    display: block;
    position: fixed;

    animation: ${({ $collapsed }) => $collapsed ? backdropFadeOut : backdropFadeIn} 200ms linear both;

    background-color: rgba(0, 0, 0, 0.5);

    top: 0;
    left: 0;
    width: 100vw;
    height: 0;


    @media only screen and (max-width: 425px) {
        height: 100vh;
    }
`