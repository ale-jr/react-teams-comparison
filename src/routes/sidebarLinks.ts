import { SidebarLink } from "interfaces/sidebarLinks";
import { faHome, faChartLine, faComment, faBullseye } from "@fortawesome/free-solid-svg-icons"
const links: SidebarLink[] = [
    {
        title: "Home",
        to: "/",
        icon: faHome,
    }, {
        title: "Analysis",
        to: "/analysis",
        icon: faChartLine,
        links: [
            {
                title: "Survey report",
                to: "/survey-report",
            },
            {
                title: "Comparison",
                to: "/comparison"
            },
            {
                title: "Custom polls",
                to: "/custom-polls"
            }
        ]
    },
    {
        title: "Feedback",
        to: "/feedback",
        icon: faComment
    },
    {
        title: "Goals",
        to: "/goals",
        icon: faBullseye
    }
]

export default links