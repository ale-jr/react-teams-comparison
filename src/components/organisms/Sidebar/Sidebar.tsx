import { SidebarLink as SidebarLinkType } from "interfaces/sidebarLinks";
import { Fragment } from "react";
import { SidebarContainer, Backdrop } from "./style";
import SidebarLink from "components/molecules/SidebarLink";

interface SidebarProps {
  activeRoute: string;
  collapsed: boolean;
  preventAnimation: boolean;
  links: SidebarLinkType[];
  onClose: () => void;
}

const renderLinks = (
  links: SidebarLinkType[],
  activeRoute: string,
  level: number = 0
): JSX.Element | JSX.Element[] => {
  return links.map(({ links: subLinks, title, to, icon }) => (
    <Fragment key={to}>
      <SidebarLink
        title={title}
        to={to}
        active={to === activeRoute}
        icon={icon}
        level={level}
      />
      {subLinks && <ul>{renderLinks(subLinks, activeRoute, level + 1)}</ul>}
    </Fragment>
  ));
};

const Sidebar: React.FC<SidebarProps> = ({
  activeRoute,
  links,
  collapsed,
  preventAnimation,
  onClose,
}) => {
  return (
    <>
      <SidebarContainer
        $collapsed={collapsed}
        $preventAnimation={preventAnimation}
      >
        <ul>{renderLinks(links, activeRoute)}</ul>
      </SidebarContainer>
      <Backdrop
        onClick={onClose}
        $collapsed={collapsed}
        $preventAnimation={preventAnimation}
      />
    </>
  );
};

export default Sidebar;
