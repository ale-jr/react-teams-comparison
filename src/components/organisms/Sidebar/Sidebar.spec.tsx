import "jest-styled-components";
import { render } from "testUtils";

import { faHome } from "@fortawesome/free-solid-svg-icons";
import { SidebarLink } from "interfaces/sidebarLinks";
import Sidebar from "./Sidebar";

describe("Sidebar", () => {
  /*
        Big snapshot tests usually are not a great idea, but due to my available time to this test
        I prefer to create some snapshots and describe what I would do in "real life":
        * Create a provider with some routes and fire click events to the links
        to confirm that active route state is working as intended;
        * Assert links and icons by label text (ensures good acessibility ðŸ˜„)
        * Create some edge cases for the sublinks recursion (empty arrays, broken links, etc)
    */

  const links: SidebarLink[] = [
    {
      icon: faHome,
      to: "/",
      title: "Home",
    },
    {
      icon: faHome,
      to: "/analysis",
      title: "Analysis",
      links: [
        {
          title: "Survey Report",
          to: "/report",
          links: [
            {
              title: "deep",
              to: "/deep",
            },
          ],
        },
        {
          title: "Comparison",
          to: "/comp",
        },
      ],
    },
    {
      to: "/whatever",
      title: "whatever",
    },
  ];

  it("Should match snapshot", () => {
    const { asFragment } = render(
      <Sidebar
        collapsed={false}
        links={links}
        activeRoute="/report"
        onClose={() => {}}
        preventAnimation={false}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
