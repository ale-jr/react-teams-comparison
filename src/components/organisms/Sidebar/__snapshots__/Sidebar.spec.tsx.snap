// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar Should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-animation: enPxJo 200ms linear both;
  animation: enPxJo 200ms linear both;
  height: 100%;
  border-right: 1px solid #F1F1F5;
  padding-top: 0.5em;
  background-color: #FFFFFF;
}

.c0 ul {
  padding: 0;
}

.c0 ul:first-child {
  -webkit-animation: chrjje 200ms linear both;
  animation: chrjje 200ms linear both;
  position: absolute;
}

.c1 {
  list-style-type: none;
}

.c2 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all 120ms linear;
  transition: all 120ms linear;
  padding: 0.5em 2em;
  padding-left: calc(1em * 1);
  color: #62656C;
  font-weight: 400;
  font-size: 1em;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
}

.c2:hover {
  color: #2678FF;
}

.c2 span.title {
  padding-left: 1em;
}

.c2 svg.icon {
  font-size: 1.5em;
}

.c2::before {
  -webkit-animation: dcGRDW 125ms linear both;
  animation: dcGRDW 125ms linear both;
  content: "";
  position: absolute;
  left: 0;
  top: 0;
  height: 100%;
  width: 4px;
  background-color: #2678FF;
}

.c3 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all 120ms linear;
  transition: all 120ms linear;
  padding: 0.5em 2em;
  padding-left: calc(1.25em + calc(1em * 2));
  color: #1C1E2A;
  font-weight: 700;
  font-size: 1em;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
}

.c3:hover {
  color: #2678FF;
}

.c3 span.title {
  padding-left: 1em;
}

.c3 svg.icon {
  font-size: 1.5em;
}

.c3::before {
  -webkit-animation: hTmvpY 125ms linear both;
  animation: hTmvpY 125ms linear both;
  content: "";
  position: absolute;
  left: 0;
  top: 0;
  height: 100%;
  width: 4px;
  background-color: #2678FF;
}

.c4 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all 120ms linear;
  transition: all 120ms linear;
  padding: 0.5em 2em;
  padding-left: calc(1.25em + calc(1em * 3));
  color: #62656C;
  font-weight: 400;
  font-size: 1em;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
}

.c4:hover {
  color: #2678FF;
}

.c4 span.title {
  padding-left: 1em;
}

.c4 svg.icon {
  font-size: 1.5em;
}

.c4::before {
  -webkit-animation: dcGRDW 125ms linear both;
  animation: dcGRDW 125ms linear both;
  content: "";
  position: absolute;
  left: 0;
  top: 0;
  height: 100%;
  width: 4px;
  background-color: #2678FF;
}

.c5 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all 120ms linear;
  transition: all 120ms linear;
  padding: 0.5em 2em;
  padding-left: calc(1.25em + calc(1em * 2));
  color: #62656C;
  font-weight: 400;
  font-size: 1em;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
}

.c5:hover {
  color: #2678FF;
}

.c5 span.title {
  padding-left: 1em;
}

.c5 svg.icon {
  font-size: 1.5em;
}

.c5::before {
  -webkit-animation: dcGRDW 125ms linear both;
  animation: dcGRDW 125ms linear both;
  content: "";
  position: absolute;
  left: 0;
  top: 0;
  height: 100%;
  width: 4px;
  background-color: #2678FF;
}

.c6 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all 120ms linear;
  transition: all 120ms linear;
  padding: 0.5em 2em;
  padding-left: calc(1.25em + calc(1em * 1));
  color: #62656C;
  font-weight: 400;
  font-size: 1em;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
}

.c6:hover {
  color: #2678FF;
}

.c6 span.title {
  padding-left: 1em;
}

.c6 svg.icon {
  font-size: 1.5em;
}

.c6::before {
  -webkit-animation: dcGRDW 125ms linear both;
  animation: dcGRDW 125ms linear both;
  content: "";
  position: absolute;
  left: 0;
  top: 0;
  height: 100%;
  width: 4px;
  background-color: #2678FF;
}

@media only screen and (max-width:425px) {
  .c0 {
    position: absolute;
    z-index: 98;
    height: calc(100% - 4.875rem);
  }
}

@media only screen and (max-width:425px) {

}

<nav
    class="c0"
  >
    <ul>
      <li
        class="c1"
      >
        <a
          class="c2"
          href="/"
          level="0"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-house fa-fw icon"
            data-icon="house"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M575.8 255.5C575.8 273.5 560.8 287.6 543.8 287.6H511.8L512.5 447.7C512.5 450.5 512.3 453.1 512 455.8V472C512 494.1 494.1 512 472 512H456C454.9 512 453.8 511.1 452.7 511.9C451.3 511.1 449.9 512 448.5 512H392C369.9 512 352 494.1 352 472V384C352 366.3 337.7 352 320 352H256C238.3 352 224 366.3 224 384V472C224 494.1 206.1 512 184 512H128.1C126.6 512 125.1 511.9 123.6 511.8C122.4 511.9 121.2 512 120 512H104C81.91 512 64 494.1 64 472V360C64 359.1 64.03 358.1 64.09 357.2V287.6H32.05C14.02 287.6 0 273.5 0 255.5C0 246.5 3.004 238.5 10.01 231.5L266.4 8.016C273.4 1.002 281.4 0 288.4 0C295.4 0 303.4 2.004 309.5 7.014L564.8 231.5C572.8 238.5 576.9 246.5 575.8 255.5L575.8 255.5z"
              fill="currentColor"
            />
          </svg>
          <span
            class="title"
          >
            Home
          </span>
        </a>
      </li>
      <li
        class="c1"
      >
        <a
          class="c2"
          href="/analysis"
          level="0"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-house fa-fw icon"
            data-icon="house"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M575.8 255.5C575.8 273.5 560.8 287.6 543.8 287.6H511.8L512.5 447.7C512.5 450.5 512.3 453.1 512 455.8V472C512 494.1 494.1 512 472 512H456C454.9 512 453.8 511.1 452.7 511.9C451.3 511.1 449.9 512 448.5 512H392C369.9 512 352 494.1 352 472V384C352 366.3 337.7 352 320 352H256C238.3 352 224 366.3 224 384V472C224 494.1 206.1 512 184 512H128.1C126.6 512 125.1 511.9 123.6 511.8C122.4 511.9 121.2 512 120 512H104C81.91 512 64 494.1 64 472V360C64 359.1 64.03 358.1 64.09 357.2V287.6H32.05C14.02 287.6 0 273.5 0 255.5C0 246.5 3.004 238.5 10.01 231.5L266.4 8.016C273.4 1.002 281.4 0 288.4 0C295.4 0 303.4 2.004 309.5 7.014L564.8 231.5C572.8 238.5 576.9 246.5 575.8 255.5L575.8 255.5z"
              fill="currentColor"
            />
          </svg>
          <span
            class="title"
          >
            Analysis
          </span>
        </a>
      </li>
      <ul>
        <li
          class="c1"
        >
          <a
            class="c3"
            href="/report"
            level="1"
          >
            <span
              class="title"
            >
              Survey Report
            </span>
          </a>
        </li>
        <ul>
          <li
            class="c1"
          >
            <a
              class="c4"
              href="/deep"
              level="2"
            >
              <span
                class="title"
              >
                deep
              </span>
            </a>
          </li>
        </ul>
        <li
          class="c1"
        >
          <a
            class="c5"
            href="/comp"
            level="1"
          >
            <span
              class="title"
            >
              Comparison
            </span>
          </a>
        </li>
      </ul>
      <li
        class="c1"
      >
        <a
          class="c6"
          href="/whatever"
          level="0"
        >
          <span
            class="title"
          >
            whatever
          </span>
        </a>
      </li>
    </ul>
  </nav>
  .c0 {
  display: block;
  position: fixed;
  z-index: 97;
  -webkit-animation: cjVSDq 200ms linear both;
  animation: cjVSDq 200ms linear both;
  background-color: rgba(0,0,0,0.5);
  top: 0;
  left: 0;
  width: 100vw;
  height: 0;
}

@media only screen and (max-width:425px) {

}

@media only screen and (max-width:425px) {
  .c0 {
    height: 100vh;
  }
}

<div
    class="c0"
  />
</DocumentFragment>
`;
